lib/tether/toolbox.go

```diff
@@ -237,3 +270,161 @@ func (t *Toolbox) halt() error {
+		spec, err := archive.DecodeFilterSpec(op, &filterSpec)
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+		diskPath, err := mountDiskLabel(diskLabel)
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+		defer unmount(op, diskPath)
+
+		// no need to join on u.Path here. u.Path == spec.Rebase, but
+		// Unpack will rebase tar headers for us. :thumbsup:
+		err = archive.Unpack(op, tr, spec, diskPath)
+		if err != nil {
+			op.Debugf(err.Error())
+		}
+		op.Debugf("Finished reading from tar archive to path %s: %s", u.Path, u.String())
+		return err
+	}
+	return defaultArchiveHandler.Read(u, tr)
+
+}
+
+func toolboxOverrideArchiveWrite(u *url.URL, tw *tar.Writer) error {
+
+	// special behavior when using disk-labels and filterspec
+	diskLabel := u.Query().Get(vsphere.DiskLabelQueryName)
+	filterSpec := u.Query().Get(vsphere.FilterSpecQueryName)
+
+	skiprecurse, _ := strconv.ParseBool(u.Query().Get(vsphere.SkipRecurseQueryName))
+	skipdata, _ := strconv.ParseBool(u.Query().Get(vsphere.SkipDataQueryName))
+
+	if diskLabel != "" && filterSpec != "" {
+		op := trace.NewOperation(context.Background(), "ToolboxOnlineDataSource: %s", u.String())
+		op.Debugf("Writing to archive from %s: %s", u.Path, u.String())
+
+		spec, err := archive.DecodeFilterSpec(op, &filterSpec)
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+
+		// get the container fs mount
+		diskPath, err := mountDiskLabel(diskLabel)
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+		defer unmount(op, diskPath)
+
+		var rc io.ReadCloser
+		if skiprecurse {
+			// we only want a single file - this is a hack while we're abusing Diff, but
+			// accomplish this by generating a single entry ChangeSet
+			changes := []dar.Change{
+				{
+					Kind: dar.ChangeModify,
+					Path: u.Path,
+				},
+			}
+
+			rc, err = archive.Tar(op, diskPath, changes, spec, !skipdata, false)
+		} else {
+			rc, err = archive.Diff(op, diskPath, "", spec, !skipdata, false)
+		}
+
+		if err != nil {
+			op.Debugf(err.Error())
+			return err
+		}
+
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer rc.Close()
+			defer wg.Done()
+			tr := tar.NewReader(rc)
+			for {
+				hdr, err := tr.Next()
+				if err == io.EOF {
+					err = nil
+					op.Debugf("Finished writing to archive from %s: %s with error %#v", u.Path, u.String(), err)
+					break
+				}
+				if err != nil {
+					op.Debugf("error writing tar: %s", err.Error())
+					break
+				}
+				op.Debugf("Writing header: %#s", *hdr)
+				err = tw.WriteHeader(hdr)
+				if err != nil {
+					op.Debugf("error writing tar header: %s", err.Error())
+					break
+				}
+				_, err = io.Copy(tw, tr)
+				if err != nil {
+					op.Debugf("error writing tar contents: %s", err.Error())
+					break
+				}
+			}
+		}()
+		wg.Wait()
+		return err
+	}
+	return defaultArchiveHandler.Write(u, tw)
+}
+
+func mountDiskLabel(label string) (string, error) {
```

Comment:
Should really be part of the Operations interface in ops_linux.go as this will not function for non-linux containers. Might involve up-leveling the function to CopyDontCrossMounts for example.
Doesn't block PR.